Add list alerts command

stephanLandwehr opened the pull request:
A user wants to know the alerts that happened in a given time frame.

Refactor list on-calls to make the user configurable.

Add code to list all alerts in a given time frame for a user.

stephanLandwehr committed:
Add list alerts command

A user wants to know the alerts that happened in a given time frame.

Refactor list on-calls to make the user configurable.

Add code to list all alerts in a given time frame for a user.

HeavyWombat commented:
This should fix #5

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,262 @@
+			return err
+		}
+
+		id, err := cmd.Flags().GetString("id")
+		if err != nil {
+			return err
+		}
```
My bad, we never touched on this topic before. There are multiple ways to do flags in Go CLIs. Since we use `cobra`, I would ask you to migrate the flags to the `cobra` style: https://github.com/spf13/cobra#working-with-flags

We should do a pairing session on this.

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,262 @@
+			user, err = client.GetUser(id, pagerduty.GetUserOptions{})
+		}
+
+		start, err := cmd.Flags().GetString("from")
+		if err != nil {
+			return err
+		}
+		end, err := cmd.Flags().GetString("to")
+		if err != nil {
+			return err
```
Same as the other flag, please use `cobra` flags instead.

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,262 @@
+		if err != nil {
+			return err
+		}
+		startTime, err := time.Parse("2006-01-02T15:04:05Z", start)
```
There are seven occurrences of `time.Parse("2006-01-02T15:04:05Z" ...`, so I would suggest to introduce a new function for this. Since the values from from the PagerDuty backend, we can even assume that the time strings are parse-able all the time with a very high confidence.

```go
func mustParsePagerDutyRFC3339ishTime(input string) time.Time {
  time, err := time.Parse("2006-01-02T15:04:05Z", input)
  if err != nil {
    panic(err)
  }

  return time
}
```

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,262 @@
+	if err != nil {
+		return false, err
+	}
+	if start.Before(end) {
+		return !time.Before(start) && !time.After(end), nil
+	}
+	if start.Equal(end) {
+		return time.Equal(start), nil
+	}
+	return !start.After(time) || !end.Before(time), nil
```
Comments would help to understand the ideas behind the three possible exits more quickly.

stephanLandwehr committed:
Refactor list-alerts command

Introduce struct for command-flags

Add function mustParsePagerDutyRFC3339ishTime

stephanLandwehr committed:
Refactor filterIncidentsByNameInLogEntries function

Increase efficiency drastically of filterIncidentsByNameInLogEntries by using goroutines.

Remove second function to filter incidents.

Fix incident-links not working.

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,230 @@
+	loadLogEntries := func(position int, wg *sync.WaitGroup) {
+		logEntry, _ := client.ListIncidentLogEntries(incidents[position].Id, pagerduty.ListIncidentLogEntriesOptions{})
+		logEntries[position] = *logEntry
+		wg.Done()
+	}
+	var wg *sync.WaitGroup
+	if numIncidents <= 50 { // if you sent 100 requests at once, you sometimes get an error. 50 requests at once is pretty safe.
+		wg = new(sync.WaitGroup)
+		wg.Add(numIncidents)
+
+		for i := 0; i < numIncidents; i++ {
+			go loadLogEntries(i, wg)
+		}
+	} else {
+		wg = new(sync.WaitGroup)
+		wg.Add(int(numIncidents / 2))
+
+		for i := 0; i < int(numIncidents/2); i++ {
+			go loadLogEntries(i, wg)
+		}
+
+		wg.Wait()
+		wg.Add(int(numIncidents / 2))
+
+		for i := int(numIncidents / 2); i < numIncidents; i++ {
+			go loadLogEntries(i, wg)
+		}
+	}
+	wg.Wait()
+
+	for i, incident := range incidents {
+		for _, logEntry := range logEntries[i].LogEntries {
+			if strings.Contains(logEntry.CommonLogEntryField.Summary, username) {
+				filteredIncidents = append(filteredIncidents, incident)
+				break
+			}
+		}
+	}
+	return filteredIncidents, nil
+}
```
Very good that you figured out that you have to rate limit the requests when doing things in parallel. There is a code pattern that solves the same thing with a bit less code and `if` blocks. Have a look at my attached suggestion. What it does is to spin up _n_  worker Go routines that then wait for work tasks to be done. Using a Go channel you can supply work tasks to the workers. A worker finishes when the work task channel is closed. That way you can relatively elegant solve the problem to not overload the back-end.

```suggestion
func filterIncidentsByNameInLogEntries(incidents []pagerduty.Incident, username string, client *pagerduty.Client) ([]pagerduty.Incident, error) {
	const parallel = 10

	type in struct {
		index    int
		incident pagerduty.Incident
	}

	var (
		errors     = []error{}
		tasks      = make(chan in, len(incidents))
		logEntries = make([]pagerduty.ListIncidentLogEntriesResponse, len(incidents))
	)

	// Fill the task channel with work to be done
	for idx, incident := range incidents {
		tasks <- in{idx, incident}
	}

	// Start a worker group to process work tasks
	var wg sync.WaitGroup
	wg.Add(parallel)
	for i := 0; i < parallel; i++ {
		go func() {
			defer wg.Done()
			for task := range tasks {
				resp, err := client.ListIncidentLogEntries(task.incident.Id, pagerduty.ListIncidentLogEntriesOptions{})
				if err != nil {
					errors = append(errors, err)
					break
				}

				logEntries[task.index] = *resp
			}
		}()
	}

	close(tasks)
	wg.Wait()

	if len(errors) > 0 {
		return nil, wrap.Errors(errors, "failed to filter incidents by name")
	}

	var filteredIncidents []pagerduty.Incident
	for i, incident := range incidents {
		for _, logEntry := range logEntries[i].LogEntries {
			if strings.Contains(logEntry.CommonLogEntryField.Summary, username) {
				filteredIncidents = append(filteredIncidents, incident)
				break
			}
		}
	}

	return filteredIncidents, nil
}
```

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+func mustParsePagerDutyRFC3339ishTime(input string) time.Time {
+	time, err := time.Parse("2006-01-02T15:04:05Z", input)
```
```suggestion
	time, err := time.Parse(time.RFC3339, input)
```

HeavyWombat reviewed (changes requested):
Sorry it took me so long to have a look at your work here. I rewrote the Go routine section, because I had to look up the finer details of the code myself first. We can have a pairing session on this part if you want.

HeavyWombat commented on internal/cmd/list-alerts.go:
```diff
@@ -0,0 +1,230 @@
+					bunt.Printf("   %s\n", incident.Description)
+				}
+
+				bunt.Printf("   *Link:* %s\n", strings.Replace(incident.Self, "api", "ibm", 1))
```
Let it show the URL as it is, even if it does not work. We figure out a solution later.
```suggestion
				bunt.Printf("   *Link:* CornflowerBlue{~%s~}\n", incident.Self)
```

stephanLandwehr committed:
Apply suggestions from code review

Refactor filterIncidentsByNameInLogEntries function.

stephanLandwehr committed:
Fix bugs

Fix command not working if the PagerDuty-account was not assigned to a team.

Fix description and notes not displaying correctly.

Fix links not working

HeavyWombat committed:
Merge branch 'main' into add/list-alerts
